src/bin/rustfmt.rs

```diff
@@ -77,6 +77,19 @@ fn lookup_and_read_project_file(input_file: &Path) -> io::Result<(PathBuf, Strin
     Ok((path, toml))
 }
 
+/// read the given config file path recursively if present else read the project file path
+fn match_config_path_or_input_file(input_file: &Path,
+                                   config_path: Option<&String>)
+                                   -> io::Result<(PathBuf, String)> {
+    match config_path {
+        Some(config_file) => {
+            lookup_and_read_project_file(config_file.as_ref())
+                .or(lookup_and_read_project_file(input_file))
```

Comment:
You are completely right that the semantics here are correct. It's just that the "backup value" is always evaluated, even when it is not necessary. It's not terribly important here, but since some IO may be involved, it's probably better to not do it if it can be skipped. The closure approach only evaluates the backup value when the original value is an error. Does that make sense?